src/agent/onefuzz/src/az_copy.rs

```diff
@@ -130,13 +131,50 @@ fn should_always_retry(err: &anyhow::Error) -> bool {
     false
 }
 
+fn file_change_detected(
+    file: &OsStr,
+    pre_copy: Result<Metadata, std::io::Error>,
+    post_copy: Result<Metadata, std::io::Error>,
+) -> bool {
+    match (pre_copy, post_copy) {
+        (Ok(pre_meta), Ok(post_meta)) => {
+            let pre_len = pre_meta.len();
+            let post_len = post_meta.len();
+            if pre_meta.len() != post_meta.len() {
+                info!(
+                    "File length changed from {} to {} for file {:?}",
+                    pre_len, post_len, file
+                );
+                true
+            } else {
+                match (pre_meta.modified(), post_meta.modified()) {
+                    (Ok(pre_modified), Ok(post_modified)) => {
+                        if pre_modified != post_modified {
+                            info!(
+                                "File modified timestamp changed from {:?} to {:?} for file {:?}",
+                                pre_modified, post_modified, file
+                            );
+                            true
```

Comment:
isn't this enough to check that the file is modified. Why do we need to check the length first ?
it seems to me that the whole function could just be 

```rust
match (pre_meta.modified(), post_meta.modified()) {
    (Ok(pre_modified), Ok(post_modified)) => pre_modified != post_modified, // + logging
    _ => false
}

```

Reply:
It depends how you copy.

Here is my F# test function:


```

let copy( n ) =
    files
    |> Seq.map (fun _ ->
        let i = System.Random().Next(Seq.length files)
        Seq.item i files
    ) |> Seq.map(fun file ->
        let path = @"C:\Users\statis\Documents\repos\onefuzz\az-copy-test\test-files\src.tar"
        System.IO.File.Copy(file.FullName, path, true)
        let fi = System.IO.FileInfo(path)
       
       //need to uncomment this to update last modified time
       //fi.LastWriteTimeUtc <- System.DateTime.UtcNow
       
        printfn "Last modified time: %A" fi.LastWriteTimeUtc

        sprintf "%s last access time: %A and size: %A" file.FullName fi.LastWriteTime ((float fi.Length)/(1024.0*1024.0))
    )

Prints:

Last modified time: 11/1/2021 9:20:10 PM
Last modified time: 11/1/2021 9:58:08 PM
Last modified time: 11/1/2021 9:18:47 PM
Last modified time: 11/1/2021 9:20:24 PM
Last modified time: 11/1/2021 9:18:27 PM
Last modified time: 11/1/2021 9:57:15 PM
Last modified time: 11/1/2021 9:20:47 PM
Last modified time: 11/1/2021 9:20:24 PM
Last modified time: 11/1/2021 9:58:08 PM
Last modified time: 11/1/2021 9:20:00 PM
Last modified time: 11/1/2021 9:18:03 PM
Last modified time: 11/1/2021 9:58:08 PM

```

I created files on 11/1/2021 and I'm running it on 11/2/2021

Reply:
not sure i get the point here. i see different value for last modified how does that show that it is not enough to check the modified value to determine if there was a change ?

Reply:
This is 12 different files being copied over and over a file named src.tar

and I created those files manually on 11/1, after a copy src.tar still has 11/1 timestamp. Even though copy is actually happening on 11/2. 

Which means that checking time stamp itself might not be sufficient.

Reply:
ok but i thought we were comparing the timestamp before and after the copy ?
so what is important here is that LastWriteTimeUtc before the copy is different from LastWriteTimeUtc after the copy ( if we are not copying the exact same file)
i think that what is happening here is that copy is just moving the file with all its metadata.

Reply:
also i think you should test it with rust. I am not sure we can assume that the copy implementation is the same on both language

Reply:
F# script is calling binary produced with Rust compiler. To test - I build Rust az-copy code into a separate executable. And I'm using F# to drive test scenario.